src/Microsoft.OData.Core/Json/JsonValueUtils.cs

```diff
@@ -335,42 +337,95 @@ internal static void WriteValue(TextWriter writer, byte[] value)
+                Debug.Assert(firstIndex < inputStringLength, "First index of the special character should be within the string");
+                buffer = BufferUtils.InitializeBufferIfRequired(buffer);
+                int bufferLength = buffer.Length;
+                int bufferIndex = 0;
+                int currentIndex = 0;
+
+                // Let's copy and flush strings up to the first index of the special char
+                while (currentIndex < firstIndex)
                 {
-                    continue;
+                    int subStrLength = firstIndex - currentIndex;
+
+                    Debug.Assert(subStrLength > 0, "SubStrLength should be greater than 0 always");
+
+                    // If the first index of the special character is larger than the buffer length,
+                    // flush everything to the buffer first and reset the buffer to the next chunk.
+                    // Otherwise copy to the buffer and go on from there.
+                    if (subStrLength >= bufferLength)
+                    {
+                        inputString.CopyTo(currentIndex, buffer, 0, bufferLength);
+                        writer.Write(buffer, 0, bufferLength);
+                        currentIndex += bufferLength;
+                    }
+                    else
+                    {
+                        inputString.CopyTo(currentIndex, buffer, 0, subStrLength);
+                        bufferIndex = subStrLength;
+                        currentIndex += subStrLength;
+                    }
                 }
 
-                // Flush out the unescaped characters we've built so far.
-                subStrLength = currentIndex - startIndex;
-                if (subStrLength > 0)
+                for (; currentIndex < inputStringLength; currentIndex++)
                 {
-                    writer.Write(inputString.Substring(startIndex, subStrLength));
-                }
+                    char c = inputString[currentIndex];
+                    string escapedString = JsonValueUtils.SpecialCharToEscapedStringMap[c];
 
-                writer.Write(SpecialCharToEscapedStringMap[c]);
-                startIndex = currentIndex + 1;
-            }
+                    // Append the unhandled characters (that do not require special treament)
+                    // to the buffer.
+                    if (escapedString == null)
+                    {
+                        buffer[bufferIndex] = c;
+                        bufferIndex++;
+                    }
+                    else
+                    {
+                        // Okay, an unhandled character was deteced.
+                        // First lets check if we can fit it in the existing buffer, if not,
+                        // flush the current buffer and reset. Add the escaped string to the buffer
+                        // and continue.
+                        int escapedStringLength = escapedString.Length;

```

Comment:
We currently don't have any escaped strings that exceed the buffer length (and I don't anticipate that we do), but in the slim chance that our assumption is not valid in the future, perhaps we can add an assert here (something like escapedStringLength >= bufferLength). Otherwise, we walk into unknown behavior from writing past the buffer boundary.

Reply:
Resolved